internal/resourcepolicies/resource_policies.go

```diff
@@ -122,11 +135,14 @@ func (p *Policies) match(res *structuredVolume) *Action {
 	return nil
 }
 
-func (p *Policies) GetMatchAction(res any) (*Action, error) {
+func (p *Policies) GetMatchAction(res any, client crclient.Client) (*Action, error) {
```

Comment:
We had an internal discussion. We feel it is not a good idea to include a dynamic retrieve into the resource policy filtering module, original it only does static match comparisons --- from a static object data against the static policy conditions. Reasons:

- The resource policy filtering is a very fundermental module, it may be used by any part, even 3rd plugins. And the API server connection may not be possible
- We cannot always assume that the object we are checking is from the current API server, e.g., if we use the filter in restore, the PVC should be from the backed up data, instead of the cluster

Reply:
Therefore, we suggest to check the PVC labels from the PVC object, instead of PV object. The existing checks on PV object stays as is.
Let us know if that works.

Reply:
@Lyndon-Li Everything else is checked on PV (size, nfs, etc). This was added to look up the PVC for the PV. The problem is that we expect that users will generally have labels on PVCs but not on PVs since the PVCs are created explicitly, while the PVs are provisioned for them. Not only that, non-privileged users aren't even allowed to edit PVs -- so if the label selector is applied to the PVs instead of the PVCs, then the usefulness of this feature is significantly reduced.

Also, these policies don't really make sense on restore -- we're deciding how to back up the volumes (fs-backup vs snapshot vs. no data backup) -- at restore, we're stuck with the backup-time decisions and can only restore via the backup method.

Reply:
@Lyndon-Li Talking to @shubham-pampattiwar just now, I think we found another way to do this that doesn't require passing in a client. In fact, in most cases we already have a PVC obj in memory that we can use.

Reply:
@Lyndon-Li @sseago Updated the PR, removed pvc fetch from volume policy parsing logic. PTAL !

Reply:
@shubham-pampattiwar @sseago
We still think it is not perfect to have the interface as `res` and `secondaryRes`. Instead, if we think none of the k8s built-in resource could represent the object to be checked against the resource policy, we should define a compound struct to include the necessary information from the both resources, e.g., PV and PVC.
This will be friendly for extensibility of resource policy.
And we also think that this will help to eliminate the differences between `case *v1.PersistentVolume` and `case *v1.Volume`, that is, one "volume" struct and represent the volume from all backup methods, i.e., PVB/PVR, CSI, etc .

Reply:
@Lyndon-Li Discussed this with @sseago what if just introduce a new struct and pass that on the `GetMatchAction` function ?
```go
type VolumeFilterData struct {
   pv PersistentVolume
   vol Volume
   pvc PersistentVolumeClaim
}
```

Reply:
First, we agree on introducing a new struct for `GetMatchAction`.
Secondly, in order to make it flexible, it's better to pick the required fields from `Volume`, `PersistentVolume` and `PersistentVolumeClaim` and add them to this new struct. It is not necessary to couple with the aforementioned k8s objects, we don't need all the fields from them.

Reply:
I get point that the upper level of the engine is to filter k8s resources, so it is reasonable to pass k8s resources. I agree on that. I also agree that based on the current code, it will be a duplication of we add another struct very similar to `structuredVolume`.
But I also feel that we should not have expected that the engine needs to filter multiple objects and filter them separately may not be reasonable. So how to create a logical struct to connect/represent these objects (not only for volume, but for all objects) may be something we can do better.

Right now, let's go with your proposal https://github.com/vmware-tanzu/velero/pull/8713#discussion_r1968117213.

Reply:
@Lyndon-Li @sseago PTAL, Thank you !